src/miner.h

```diff
@@ -151,7 +151,10 @@ int64_t UpdateTime(CBlockHeader *pblock, const Consensus::Params &consensusParam
 UniValue SubmitBlock(CBlock &block);
 /** Make a block template to send to miners. */
 // implemented in mining.cpp
-UniValue mkblocktemplate(const UniValue &params, int64_t coinbaseSize = -1, CBlock *pblockOut = nullptr);
+UniValue mkblocktemplate(const UniValue &params,
+    int64_t coinbaseSize = -1,
+    CBlock *pblockOut = nullptr,
+    boost::shared_ptr<CReserveScript> coinbaseScript = boost::shared_ptr<CReserveScript>());
```

Comment:
why not using `std::shared_ptr`? (also on other parts of the PR)

Reply:
Yeah -- I'm not clear on if I made a good decision here.  In fact reading the code again it's hard to defend this decision.

I think I saw other parts of the code do this -- namely the wallet code expected a `boost::shared_ptr<CReserveScript>` and I wanted to be "in line" with that -- but as I read the code it's hard to justify it.

A simple reference to a `CReserveScript` or even a `CScript` would suffice, really.  It's not like that data structure is expensive to copy or this will be called thousands of times per second (and if it were it's arguable the `prevector`-based `CReserveScript` or `CScript` directly is faster/less heap intense).

So.. actually just passing a reference to the `CReserveScript` should be fine... and caching it in the `static` works.

Should I go ahead and do that?

--

Now.. for that matter -- if I may ask -- why does the `GetScriptForMining` call in the wallet expect a `shared_ptr` and not a simple reference? (Just curious).

Reply:
Oh -- I know why I did it. In `mining.cpp generateBlocks` ... it actually also takes a `boost::shared_ptr<CReserveScript>` and without thinking I figured there's something about doing it that way.

If you don't mind -- and if you know -- why does that take a `shared_ptr` there?

Reply:
> Now.. for that matter -- if I may ask -- why does the `GetScriptForMining` call in the wallet expect a `shared_ptr` and not a simple reference? (Just curious).

dunno, maybe @gandrewstone 	could chime in here. 

My question about using `shared_ptr` was more related to the fact that we are trying to move away from boost (even thou slowly) and I didn't realize you chose `boost::shared_ptr` because `generateBlocks` use this type for one of its param

Reply:
Oh .. I see.. the wallet expects a pointer because it may return a polymorphic type (virtual methods to `CReserveScript`) -- which makes me wonder -- does this codepath need to call `KeepKey()` after generating a block with the wallet address in coinbase? (Note this PR did not introduce this concern -- it would have been there before this PR anyway).

Reply:
> > Now.. for that matter -- if I may ask -- why does the `GetScriptForMining` call in the wallet expect a `shared_ptr` and not a simple reference? (Just curious).
> 
> dunno, maybe @gandrewstone could chime in here.
> 
> My question about using `shared_ptr` was more related to the fact that we are trying to move away from boost (even thou slowly) and I didn't realize you chose `boost::shared_ptr` because `generateBlocks` use this type for one of its param

I agree that using the `std::` variety (which is basically boost anyway) is the way to go now that C++14, C++17, etc are here. I guess using the `boost` parts not in `std` would be ideal..

Yeah I would like to know what to do here -- I really just wanted to "behave" like the other code path -- note that privately a `boost::shared_ptr` would still be needed here to grab a public key from the wallet in the case where caller didn't pass one down....

I'll do whatever you guys say.  Any guidance is appreciated here. I'm new to the codebase (well, it's an old familiar friend as it resembles what I worked on in the past a great deal, with some minor and major variation.. :) ).

Reply:
Please use references when the function arg's life is short (as short as the function).  The reference will be protected from deletion by the caller's context.

Use the c++14 std:: variety whenever possible, and you can convert some boost:: equivalents opportunistically.  However, if enough conversions pile up to obscure the point of the PR, they should be isolated in their own PR.